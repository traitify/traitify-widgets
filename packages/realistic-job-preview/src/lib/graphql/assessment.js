export const create = `
  mutation createAssessment(
    $localeKey: String!,
    $profileID: String!,
    $surveyID: String!
  ) {
    createAssessment(
      localeKey: $localeKey,
      profileId: $profileID,
      surveyId: $surveyID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const get = `
  query assessment(
    $assessmentID: ID!
  ) {
    assessment(
      id: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const path = "/realistic-job-previews/graphql";

export const start = `
  mutation startAssessment(
    $assessmentID: ID!
  ) {
    startAssessment(
      id: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const update = `
  mutation updateAssessmentAnswer(
    $answers: [AssessmentAnswer],
    $assessmentID: String!
  ) {
    updateAssessmentAnswer(
      answers: $answers,
      assessmentId: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;
