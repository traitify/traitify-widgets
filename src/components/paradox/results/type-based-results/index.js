import PropTypes from "prop-types";
import PersonalityBase from "components/results/type-based-results/personality-base";
import PersonalityDetails from "components/results/type-based-results/personality-details";
import PersonalityTraits from "components/results/type-based-results/personality-traits";
import PersonalityTypes from "components/results/type-based-results/personality-types";
import withTraitify from "lib/with-traitify";
import style from "../style.scss";

function TypeBasedResults({setElement, ...props}) {
  return (
    <section className={style.container} ref={setElement}>
      <PersonalityBase {...props} />
      <PersonalityTypes {...props} />
      <PersonalityTraits {...props} />
      <PersonalityDetails {...props} />
    </section>
  );
}

TypeBasedResults.propTypes = {
  setElement: PropTypes.func.isRequired
};

export {TypeBasedResults as Component};
export default withTraitify(TypeBasedResults);
