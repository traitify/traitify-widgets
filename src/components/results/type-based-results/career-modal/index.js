/* eslint-disable react/no-unused-class-component-methods */
import {
  faAdjust,
  faChartBar,
  faExternalLinkSquareAlt,
  faCheckSquare,
  faDollarSign,
  faGlobeAmericas,
  faGraduationCap,
  faInfoCircle,
  faLeaf,
  faLightbulb,
  faSquare,
  faTimes,
  faQuestion
} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import {Component} from "react";
import DangerousHTML from "lib/helpers/dangerous-html";
import HighChart from "components/highchart";
import Icon from "lib/helpers/icon";
import TraitifyPropTypes from "lib/helpers/prop-types";
import withTraitify from "lib/with-traitify";
import {Component as Paradox} from "components/paradox/results/personality/career/modal";
import style from "./style.scss";

class CareerModal extends Component {
  static propTypes = {
    assessment: PropTypes.shape({
      personality_traits: PropTypes.arrayOf(
        PropTypes.shape({
          personality_trait: PropTypes.shape({
            definition: PropTypes.string.isRequired,
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired
          }).isRequired,
          score: PropTypes.number.isRequired
        })
      )
    }),
    isReady: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    ui: TraitifyPropTypes.ui.isRequired
  };
  static defaultProps = {assessment: null};
  constructor(props) {
    super(props);

    this.state = {
      career: null,
      show: false,
      showLegend: false
    };
  }
  componentDidMount() {
    this.props.ui.trigger("CareerModal.initialized", this);
    this.props.ui.on("CareerModal.career", this.setCareer);
    this.props.ui.on("CareerModal.hide", this.hide);
    this.props.ui.on("CareerModal.show", this.show);
  }
  componentDidUpdate() {
    this.props.ui.trigger("CareerModal.updated", this);
  }
  componentWillUnmount() {
    this.props.ui.off("CareerModal.career", this.setCareer);
    this.props.ui.off("CareerModal.hide", this.hide);
    this.props.ui.off("CareerModal.show", this.show);
  }
  close = () => { this.props.ui.trigger("CareerModal.hide", this); };
  hide = () => { this.setState({show: false}); };
  setCareer = () => { this.setState({career: this.props.ui.current["CareerModal.career"]}); };
  show = () => { this.setState({show: true}); };
  toggleLegend = () => { this.setState((state) => ({showLegend: !state.showLegend})); };
  render() {
    const {career, show, showLegend} = this.state;
    const {assessment, isReady, translate} = this.props;
    if(!show || !career) { return null; }
    if(!isReady("results")) { return null; }

    let salary = career.salary_projection && career.salary_projection.annual_salary_mean;
    salary = salary && `$${Math.round(salary)}`;
    let growth = career.employment_projection && career.employment_projection.percent_growth_2022;
    growth = growth && `${growth}%`;

    const careerTraitIDs = career.personality_traits.map((trait) => trait.personality_trait.id);
    const careerTraits = career.personality_traits.map((trait) => ({
      description: trait.personality_trait.definition,
      name: trait.personality_trait.name,
      y: trait.weight
    }));
    const assessmentTraits = assessment.personality_traits
      .filter((trait) => careerTraitIDs.includes(trait.personality_trait.id))
      .map((trait) => ({
        description: trait.personality_trait.definition,
        name: trait.personality_trait.name,
        y: Math.round((trait.score + 100) / 2)
      }));

    return (
      <div className={`${style.modal} ${style.container}`} role="dialog">
        <section className={style.modalContainer}>
          <Icon aria-label={translate("close")} className={style.close} icon={faTimes} onClick={this.close} tabIndex="-1" />
          <div className={style.modalContent}>
            <img className={style.image} alt={career.title} src={career.picture} />
            <h2 className={style.title}>{career.title}</h2>
            <p className={style.description}>{career.description}</p>
            <hr />
            <h3 className={style.heading}><Icon icon={faInfoCircle} /> {translate("career_information")}:</h3>
            <ul className={style.info}>
              <li>
                <h4><Icon icon={faDollarSign} /> {translate("salary_mean")}:</h4>
                <div className={`${style.infoText} ${style.salary}`}>{salary}</div>
              </li>
              <li>
                <h4><Icon icon={faChartBar} /> {translate("employment_growth")}:</h4>
                <div className={`${style.infoText} ${style.growth}`}>{growth}</div>
              </li>
              <li>
                <h4><Icon icon={faGraduationCap} /> {translate("education")}:</h4>
                <div className={`${style.infoText} ${style.education}`}>
                  {translate(`experience_level_${career.experience_level.id}`)}
                </div>
              </li>
            </ul>
            <ul className={style.info}>
              <li>
                <h4><Icon icon={faLightbulb} /> {translate("bright_future")}:</h4>
                <div className={style.infoText}>
                  <input aria-label={translate("bright_future")} checked={career.bright_outlooks.length > 0} className={style.check} disabled={true} type="checkbox" readOnly={true} />
                  <Icon icon={career.bright_outlooks.length > 0 ? faCheckSquare : faSquare} />
                </div>
              </li>
              <li>
                <h4><Icon icon={faLeaf} /> {translate("green_career")}:</h4>
                <div className={style.infoText}>
                  <input aria-label={translate("green_career")} checked={career.green_categories.length > 0} className={style.check} disabled={true} type="checkbox" readOnly={true} />
                  <Icon icon={career.green_categories.length > 0 ? faCheckSquare : faSquare} />
                </div>
              </li>
              <li>
                <div className={style.center}>
                  <button className={style.legendToggle} onClick={this.toggleLegend} title={translate("more_information")} type="button">
                    {translate("help")} <Icon icon={faQuestion} />
                  </button>
                </div>
              </li>
            </ul>
            <p className={style.center}>
              <a className={style.btnPrimary} href={`http://www.onetonline.org/link/summary/${career.id}`} target="_blank" rel="noopener noreferrer" title={translate("view_on_onet")}>{translate("view_on_onet")}</a>
            </p>
            {showLegend && (
              <div className={style.legend}>
                <ul className={style.info}>
                  <li>
                    <h4><Icon icon={faDollarSign} /> {translate("salary_mean")}:</h4>
                    <DangerousHTML html={translate("salary_mean_html")} tag="p" />
                  </li>
                  <li>
                    <h4><Icon icon={faChartBar} /> {translate("employment_growth")}:</h4>
                    <DangerousHTML html={translate("employment_growth_html")} tag="p" />
                  </li>
                  <li>
                    <h4><Icon icon={faGraduationCap} /> {translate("education")}:</h4>
                    <DangerousHTML html={translate("education_html")} tag="p" />
                  </li>
                </ul>
                <ul className={style.info}>
                  <li>
                    <h4><Icon icon={faLightbulb} /> {translate("bright_future")}:</h4>
                    <DangerousHTML html={translate("bright_future_html")} tag="p" />
                  </li>
                  <li>
                    <h4><Icon icon={faLeaf} /> {translate("green_career")}:</h4>
                    <DangerousHTML html={translate("green_career_html")} tag="p" />
                  </li>
                </ul>
                <p className={style.center}>
                  <button className={style.legendToggle} onClick={this.toggleLegend} title={translate("close")} type="button">{translate("close")}</button>
                </p>
              </div>
            )}
            <hr />
            <h3 className={style.heading}><Icon icon={faGlobeAmericas} /> {translate("education_level")}</h3>
            <ol className={style.experience}>
              {[1, 2, 3, 4, 5].map((level) => (
                <li key={level} className={career.experience_level.id >= level ? style.active : ""} />
              ))}
            </ol>
            <p>{career.experience_level.experience}</p>
            <hr />
            <h3 className={style.heading}><Icon icon={faAdjust} /> {translate("match_rate")}</h3>
            <DangerousHTML html={translate("match_rate_html", {match_rate: career.score.toFixed(1)})} tag="p" />
            <div>
              <HighChart
                type="column"
                xAxis={{type: "category"}}
                chart={{marginRight: 50, marginLeft: 50}}
                legend={{layout: "horizontal", align: "center", width: 500, itemStyle: {width: 300}, margin: 30}}
                series={[{
                  name: `${career.title} Traits`,
                  data: careerTraits.sort((a, b) => ((a.name > b.name) ? 1 : -1))
                }, {
                  name: "Your Traits",
                  data: assessmentTraits.sort((a, b) => ((a.name > b.name) ? 1 : -1))
                }]}
              />
            </div>
            <div ref={(customContent) => { this.customContent = customContent; }} />
            <hr />
            <div>
              <h3><Icon icon={faExternalLinkSquareAlt} /> {translate("job_links")}</h3>
              <ul className={style.jobsLinks}>
                <li>
                  <a className={style.jobLink} href={`http://www.mynextmove.org/profile/summary/${career.id}`} target="_blank" rel="noopener noreferrer" title="My Next Move">
                    <img alt="My Next Move" src="https://cdn.traitify.com/assets/images/job-links/my-next-move.jpg" />
                  </a>
                </li>
                <li>
                  <a className={style.jobLink} href={`http://www.onetonline.org/link/summary/${career.id}`} target="_blank" rel="noopener noreferrer" title="O*Net Online">
                    <img alt="O*Net Online" src="https://cdn.traitify.com/assets/images/job-links/o-net-2.jpg" />
                  </a>
                </li>
                <li>
                  <a className={style.jobLink} href={`http://www.miproximopaso.org/profile/summary/${career.id}`} target="_blank" rel="noopener noreferrer" title="Mi Proximo Paso">
                    <img alt="Mi Proximo Paso" src="https://cdn.traitify.com/assets/images/job-links/mi-proximo-paso.jpg" />
                  </a>
                </li>
                <li>
                  <a className={style.jobLink} href={`http://www.mynextmove.org/vets/profile/summary/${career.id}`} target="_blank" rel="noopener noreferrer" title="My Next Move for Veterans">
                    <img alt="My Next Move for Veterans" src="https://cdn.traitify.com/assets/images/job-links/my-next-move-veterans.jpg" />
                  </a>
                </li>
              </ul>
            </div>
            <hr />
          </div>
        </section>
      </div>
    );
  }
}

export {CareerModal as Component};
export default withTraitify(CareerModal, {paradox: Paradox});
