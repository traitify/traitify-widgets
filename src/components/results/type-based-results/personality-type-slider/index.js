import PropTypes from "prop-types";
import {Component} from "react";
import TraitifyPropTypes from "lib/helpers/prop-types";
import withTraitify from "lib/with-traitify";
import PersonalityTypeSlide from "../personality-type-slide";
import TypeButton from "./type-button";
import style from "./style.scss";

class PersonalityTypeSlider extends Component {
  static defaultProps = {assessment: null};
  static propTypes = {
    assessment: PropTypes.shape({
      personality_types: PropTypes.arrayOf(
        PropTypes.shape({
          personality_type: PropTypes.shape({
            id: PropTypes.string.isRequired
          }).isRequired
        }).isRequired
      )
    }),
    isReady: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    ui: TraitifyPropTypes.ui.isRequired
  };
  constructor(props) {
    super(props);

    this.state = {activeType: null};
  }
  componentDidMount() {
    this.props.ui.trigger("PersonalityTypeSlider.initialized", this);
    this.props.ui.on("Assessment.activeType", this.getActiveType);

    const activeType = this.props.ui.current["Assessment.activeType"];
    if(activeType) { this.setState({activeType}); }
  }
  componentDidUpdate() {
    this.props.ui.trigger("PersonalityTypeSlider.updated", this);
  }
  componentWillUnmount() {
    this.props.ui.off("Assessment.activeType", this.getActiveType);
  }
  getActiveType = () => {
    this.setState({activeType: this.props.ui.current["Assessment.activeType"]});
  };
  setActive = (type) => {
    this.props.ui.trigger("PersonalityTypeSlider.changeType", this, type);
    this.props.ui.trigger("Assessment.activeType", this, type);
  };
  render() {
    if(!this.props.isReady("results")) { return null; }

    const {activeType} = this.state;
    const {assessment: {personality_types: personalityTypes}, translate} = this.props;

    let backType;
    let nextType;
    if(activeType) {
      const {id} = activeType.personality_type;
      const index = personalityTypes.findIndex((type) => type.personality_type.id === id);
      backType = personalityTypes[index - 1];
      nextType = personalityTypes[index + 1];
    }

    return (
      <div className={style.slider}>
        {backType && (
          <TypeButton className={style.back} type={backType} setActive={this.setActive}>
            <img src="https://cdn.traitify.com/assets/images/js/arrow_left.svg" alt={translate("back")} />
          </TypeButton>
        )}
        <ul>
          {personalityTypes.map((type) => (
            <PersonalityTypeSlide key={type.personality_type.id} type={type} {...this.props} />
          ))}
        </ul>
        {nextType && (
          <TypeButton className={style.next} type={nextType} setActive={this.setActive}>
            <img src="https://cdn.traitify.com/assets/images/js/arrow_right.svg" alt={translate("next")} />
          </TypeButton>
        )}
      </div>
    );
  }
}

export {PersonalityTypeSlider as Component};
export default withTraitify(PersonalityTypeSlider);
