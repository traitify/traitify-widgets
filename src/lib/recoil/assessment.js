import {selector, selectorFamily} from "recoil";
import {assessmentsState} from "./assessments";
import {
  activeState,
  activeIDState,
  activeTypeState,
  cacheState,
  graphqlState,
  httpState,
  localeState,
  safeCacheKeyState
} from "./base";

// TODO: Return error instead of null for cognitive and external
// return {errors: response.errors, id};
export const cognitiveAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const params = {
      query: GraphQL.cognitive.get,
      variables: {localeKey: get(localeState), testID: id}
    };
    const response = await http.post({path: GraphQL.cognitive.path, params});
    if(response.errors) {
      console.warn("cognitive-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.cognitiveTest;
    if(!assessment?.completed) { return assessment; }

    cache.set(cacheKey, assessment);

    return assessment;
  },
  key: "assessment/cognitive"
});

export const externalAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const query = {
      params: {
        query: GraphQL.external.get,
        variables: {id}
      },
      path: GraphQL.external.path
    };
    if(http.version === "v1") { query.version = GraphQL.external.version; }

    const response = await http.post(query);
    if(response.errors) {
      console.warn("external-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.getAssessment;
    if(!assessment?.completedAt) { return assessment; }

    cache.set(cacheKey, assessment);

    return assessment;
  },
  key: "assessment/external"
});

export const personalityAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const params = {
      data: "archetype,blend,instructions,recommendation,slides,types,traits",
      locale_key: get(localeState)
    };
    const http = get(httpState);
    const response = await http.get({path: `/assessments/${id}`, params});
    if(!response?.completed_at) { return response; }

    cache.set(cacheKey, response);

    return response;
  },
  key: "assessment/personality"
});

export const genericAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached && !cached.completed) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const params = {
      query: GraphQL.generic.questions,
      variables: {assessmentID: id}
    };

    const response = await http.post({path: GraphQL.generic.path, params});
    if(response.errors) {
      console.warn("generic-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.genericSurveyQuestions;
    if(!assessment?.completedAt) { return assessment; }

    const query = GraphQL.generic.result;
    const variables = {assessmentID: id};

    const resultResponse = await http.post(GraphQL.generic.path, {query, variables});
    if(resultResponse.errors) { return null; }
    const result = resultResponse.data.genericAssessmentResult.assessment;

    cache.set(cacheKey, result);

    return result;
  },
  key: "assessment/generic"
});

export const assessmentQuery = selectorFamily({
  get: ({id, surveyType}) => async({get}) => {
    if(surveyType === "cognitive") { return get(cognitiveAssessmentQuery(id)); }
    if(surveyType === "external") { return get(externalAssessmentQuery(id)); }
    if(surveyType === "generic") { return get(genericAssessmentQuery(id)); }
    if(surveyType === "personality") { return get(personalityAssessmentQuery(id)); }

    return null;
  },
  key: "assessment"
});

export const activeAssessmentQuery = selector({
  get: async({get}) => {
    const id = get(activeIDState);
    const surveyType = get(activeTypeState);

    return get(assessmentQuery({id, surveyType}));
  },
  key: "assessment/active"
});

export const completedAssessmentQuery = selectorFamily({
  get: ({surveyType}) => async({get}) => {
    if(!surveyType) {
      const active = get(activeState);
      if(!active?.completed) { return null; }

      return get(activeAssessmentQuery);
    }

    const assessments = get(assessmentsState);
    if(!assessments) { return null; }

    const assessment = assessments
      .find(({completed, surveyType: type}) => type === surveyType && completed);
    if(!assessment) { return null; }

    return get(assessmentQuery({id: assessment.id, surveyType}));
  },
  key: "assessment/completed"
});
