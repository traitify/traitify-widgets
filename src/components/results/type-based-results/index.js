import {
  faThLarge,
  faUser
} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import {Component as Paradox} from "components/paradox/results/type-based-results";
import Icon from "lib/helpers/icon";
import withTraitify from "lib/with-traitify";
import PersonalityBase from "./personality-base";
import PersonalityDetails from "./personality-details";
import PersonalityTraits from "./personality-traits";
import PersonalityTypes from "./personality-types";
import style from "../style.scss";

function TypeBasedResults(props) {
  const {getOption, translate} = props;
  const allowHeaders = getOption("allowHeaders");

  return (
    <section>
      {allowHeaders && (
        <p className={style.heading}><Icon icon={faUser} /> {translate("personality_blend")}</p>
      )}
      <PersonalityBase {...props} />
      {allowHeaders && (
        <p className={style.heading}><Icon icon={faThLarge} /> {translate("personality_breakdown")}</p>
      )}
      <PersonalityTypes {...props} />
      {allowHeaders && (
        <p className={style.heading}><Icon icon={faUser} /> {translate("personality_traits")}</p>
      )}
      <PersonalityTraits {...props} />
      {allowHeaders && (
        <p className={style.heading}><Icon icon={faThLarge} /> {translate("personality_details")}</p>
      )}
      <PersonalityDetails {...props} />
    </section>
  );
}

TypeBasedResults.propTypes = {
  getOption: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired
};

export default withTraitify(TypeBasedResults, {paradox: Paradox});
