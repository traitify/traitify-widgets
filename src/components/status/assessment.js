import {faCheck} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import {useSetRecoilState} from "recoil";
import Icon from "components/common/icon";
import get from "lib/common/object/get";
import useListener from "lib/hooks/use-listener";
import useOption from "lib/hooks/use-option";
import {activeState} from "lib/recoil";
import style from "./style.scss";

// TODO: Extract text to translate
const translations = {
  complete: "Complete",
  loading: "Loading", // Already translated
  status: {
    start: "Start Assessment"
  },
  survey: {
    cognitive_assessment: "Cognitive Assessment",
    external_assessment: "External Assessment",
    personality_assessment: "Personality Assessment"
  }
};

function Button({assessment}) {
  const listener = useListener();
  const options = useOption("status") || {};
  const redirect = get(options, "allowRedirect", true);
  const setActive = useSetRecoilState(activeState);

  if(assessment.completed) {
    return <button disabled={true} type="button">{translations.complete}</button>;
  }

  if(assessment.skipped) {
    return <button disabled={true} type="button">Skipped</button>;
  }

  if(assessment.link && redirect) {
    return <a href={assessment.link}>{translations.status.start}</a>;
  }

  if(assessment.loading) {
    return <button disabled={true} type="button">{translations.loading}</button>;
  }

  const start = () => {
    listener.trigger("Survey.start", {assessment});
    setActive({...assessment});
  };

  return <button onClick={start} type="button">{translations.status.start}</button>;
}

Button.propTypes = {
  assessment: PropTypes.shape({
    completed: PropTypes.bool,
    link: PropTypes.string,
    loading: PropTypes.bool,
    skipped: PropTypes.bool
  }).isRequired
};

function Assessment({assessment}) {
  const surveyName = assessment.surveyName || translations.survey[`${assessment.surveyType}_assessment`];

  return (
    <div className={[style.assessment, assessment.completed && style.inactive].filter(Boolean).join(" ")}>
      <div>
        {assessment.completed && <Icon className={style.icon} icon={faCheck} />}
        <span className={style.text}>{surveyName}</span>
      </div>
      <Button assessment={assessment} />
    </div>
  );
}

Assessment.propTypes = {
  assessment: PropTypes.shape({
    completed: PropTypes.bool,
    skipped: PropTypes.bool,
    surveyName: PropTypes.string,
    surveyType: PropTypes.string.isRequired
  }).isRequired
};

export default Assessment;
