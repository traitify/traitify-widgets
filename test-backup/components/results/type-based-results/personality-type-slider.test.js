import {Component} from "components/results/type-based-results/personality-type-slider";
import TypeButton from "components/results/type-based-results/personality-type-slider/type-button";
import ComponentHandler from "support/component-handler";
import assessment from "support/json/assessment/type-based.json";

jest.mock("components/results/type-based-results/personality-type-slide", () => ((props) => (
  <div className="mock">Slide - {props.type.personality_type.name}</div>
)));
jest.mock("lib/with-traitify", () => ((value) => value));

describe("PersonalityTypeSlider", () => {
  let props;

  beforeEach(() => {
    props = {
      assessment,
      isReady: jest.fn().mockName("isReady").mockImplementation(() => true),
      translate: jest.fn().mockName("translate").mockImplementation((value) => value),
      ui: {
        current: {},
        off: jest.fn().mockName("off"),
        on: jest.fn().mockName("on"),
        trigger: jest.fn().mockName("trigger")
      }
    };
  });

  describe("callbacks", () => {
    it("triggers initialization", () => {
      const component = new ComponentHandler(<Component {...props} />);

      expect(props.ui.trigger).toHaveBeenCalledWith("PersonalityTypeSlider.initialized", component.instance);
    });

    it("triggers update", () => {
      const component = new ComponentHandler(<Component {...props} />);
      component.updateProps();

      expect(props.ui.trigger).toHaveBeenCalledWith("PersonalityTypeSlider.updated", component.instance);
    });

    it("triggers callbacks on setActive", () => {
      const component = new ComponentHandler(<Component {...props} />);
      const type = props.assessment.personality_types[0];
      component.instance.setActive(type);

      expect(props.ui.trigger).toHaveBeenCalledWith("PersonalityTypeSlider.changeType", component.instance, type);
      expect(props.ui.trigger).toHaveBeenCalledWith("Assessment.activeType", component.instance, type);
    });
  });

  describe("active type", () => {
    it("listens to changes", () => {
      const component = new ComponentHandler(<Component {...props} />);

      expect(props.ui.on).toHaveBeenCalledWith("Assessment.activeType", component.instance.getActiveType);
    });

    it("unlistens to changes", () => {
      const component = new ComponentHandler(<Component {...props} />);
      const {getActiveType} = component.instance;
      component.unmount();

      expect(props.ui.off).toHaveBeenCalledWith("Assessment.activeType", getActiveType);
    });

    it("sets type if present", () => {
      const type = props.assessment.personality_types[0];
      props.ui.current["Assessment.activeType"] = type;
      const component = new ComponentHandler(<Component {...props} />);

      expect(component.state.activeType).toEqual(type);
    });

    it("updates state with new type", () => {
      const component = new ComponentHandler(<Component {...props} />);
      const type = props.assessment.personality_types[0];
      props.ui.current["Assessment.activeType"] = type;
      component.instance.getActiveType();

      expect(component.state.activeType).toEqual(type);
    });
  });

  describe("type button", () => {
    beforeEach(() => {
      props = {
        setActive: jest.fn().mockName("setActive"),
        type: props.assessment.personality_types[0]
      };
    });

    it("renders component", () => {
      const component = new ComponentHandler(<TypeButton {...props} />);

      expect(component.tree).toMatchSnapshot();
    });

    it("calls setActive prop with type", () => {
      const component = new ComponentHandler(<TypeButton {...props} />);
      component.instance.setActive();

      expect(props.setActive).toHaveBeenCalledWith(props.type);
    });
  });

  it("renders component", () => {
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders nothing if not ready", () => {
    props.assessment = null;
    props.isReady.mockImplementation(() => false);
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders back button", () => {
    props.ui.current["Assessment.activeType"] = props.assessment.personality_types.slice(-1)[0];
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders next button", () => {
    props.ui.current["Assessment.activeType"] = props.assessment.personality_types[0];
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders both buttons", () => {
    props.ui.current["Assessment.activeType"] = props.assessment.personality_types[1];
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });
});
