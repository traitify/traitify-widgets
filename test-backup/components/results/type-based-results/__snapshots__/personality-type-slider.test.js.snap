// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalityTypeSlider renders back button 1`] = `
<div
  className="slider"
>
  <button
    className="back"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="back"
      src="https://cdn.traitify.com/assets/images/js/arrow_left.svg"
    />
  </button>
  <ul>
    <div
      className="mock"
    >
      Slide - 
      Naturalist
    </div>
    <div
      className="mock"
    >
      Slide - 
      Analyzer
    </div>
    <div
      className="mock"
    >
      Slide - 
      Inventor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Mentor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Planner
    </div>
    <div
      className="mock"
    >
      Slide - 
      Visionary
    </div>
    <div
      className="mock"
    >
      Slide - 
      Action-Taker
    </div>
  </ul>
</div>
`;

exports[`PersonalityTypeSlider renders both buttons 1`] = `
<div
  className="slider"
>
  <button
    className="back"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="back"
      src="https://cdn.traitify.com/assets/images/js/arrow_left.svg"
    />
  </button>
  <ul>
    <div
      className="mock"
    >
      Slide - 
      Naturalist
    </div>
    <div
      className="mock"
    >
      Slide - 
      Analyzer
    </div>
    <div
      className="mock"
    >
      Slide - 
      Inventor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Mentor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Planner
    </div>
    <div
      className="mock"
    >
      Slide - 
      Visionary
    </div>
    <div
      className="mock"
    >
      Slide - 
      Action-Taker
    </div>
  </ul>
  <button
    className="next"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="next"
      src="https://cdn.traitify.com/assets/images/js/arrow_right.svg"
    />
  </button>
</div>
`;

exports[`PersonalityTypeSlider renders component 1`] = `
<div
  className="slider"
>
  <ul>
    <div
      className="mock"
    >
      Slide - 
      Naturalist
    </div>
    <div
      className="mock"
    >
      Slide - 
      Analyzer
    </div>
    <div
      className="mock"
    >
      Slide - 
      Inventor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Mentor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Planner
    </div>
    <div
      className="mock"
    >
      Slide - 
      Visionary
    </div>
    <div
      className="mock"
    >
      Slide - 
      Action-Taker
    </div>
  </ul>
</div>
`;

exports[`PersonalityTypeSlider renders next button 1`] = `
<div
  className="slider"
>
  <ul>
    <div
      className="mock"
    >
      Slide - 
      Naturalist
    </div>
    <div
      className="mock"
    >
      Slide - 
      Analyzer
    </div>
    <div
      className="mock"
    >
      Slide - 
      Inventor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Mentor
    </div>
    <div
      className="mock"
    >
      Slide - 
      Planner
    </div>
    <div
      className="mock"
    >
      Slide - 
      Visionary
    </div>
    <div
      className="mock"
    >
      Slide - 
      Action-Taker
    </div>
  </ul>
  <button
    className="next"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="next"
      src="https://cdn.traitify.com/assets/images/js/arrow_right.svg"
    />
  </button>
</div>
`;

exports[`PersonalityTypeSlider renders nothing if not ready 1`] = `null`;

exports[`PersonalityTypeSlider type button renders component 1`] = `
<button
  className={null}
  onClick={[Function]}
  type="button"
/>
`;
